pandas/io/formats/html.py

```diff
@@ -42,6 +42,15 @@ def __init__(self, formatter, classes=None, notebook=False, border=None,
         self.border = border
         self.table_id = table_id
         self.render_links = render_links
+        # see gh-22579
+        # Column misalignment also occurs for
+        # a standard index when the columns index is named.
+        # Determine if ANY column names need to be displayed
+        # since if the row index is not displayed a column of
+        # blank cells need to be included before the DataFrame values.
+        self.show_col_idx_names = all((self.fmt.has_column_names,
```

Comment:
Sorry for questions but still trying to wrap my head around implementation. Based off of the comment, why is this `all` here and not `any`? Wouldn't any of these require there to be a cell where a column index name would be placed?

Reply:
`index=False` with a single level row index and multi-level columns index with named columns but not all named...

```python
index = pd.MultiIndex.from_product([['a','b'], ['c','d'], ['e','f']], names=[
                 'foo',None, 'baz'])
df = pd.DataFrame(np.arange(64).reshape(8,8), columns=index)
result = df.to_html(max_rows=4, max_cols=4, index=False)
print(result)
```

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>foo</th>
      <th colspan="2" halign="left">a</th>
      <th>...</th>
      <th colspan="2" halign="left">b</th>
    </tr>
    <tr>
      <th colspan="2" halign="left">c</th>
      <th>...</th>
      <th colspan="2" halign="left">d</th>
    </tr>
    <tr>
      <th>baz</th>
      <th>e</th>
      <th>f</th>
      <th>...</th>
      <th>e</th>
      <th>f</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>1</td>
      <td>6</td>
      <td>7</td>
    </tr>
    <tr>
      <td>8</td>
      <td>9</td>
      <td>14</td>
      <td>15</td>
    </tr>
    <tr>
      <td>48</td>
      <td>49</td>
      <td>54</td>
      <td>55</td>
    </tr>
    <tr>
      <td>56</td>
      <td>57</td>
      <td>62</td>
      <td>63</td>
    </tr>
  </tbody>
</table>

Note: missing truncation indicators in data now fixed in master.

the misalignment of the column names is due to the logic being applied within the level generating loop..

https://github.com/pandas-dev/pandas/blob/d43ac979fdf3038c713de93c7b27b4140be6c9c5/pandas/io/formats/html.py#L270-L275

hence class-level variable needed to check if ANY names need to be displayed to determine alignment.

ALL condition is to determine in ANY names should be displayed given the `to_html` parameters and uses similar logic as `to_string` etc.

https://github.com/pandas-dev/pandas/blob/d43ac979fdf3038c713de93c7b27b4140be6c9c5/pandas/io/formats/format.py#L796-L803

and the rows in `to_html`..

https://github.com/pandas-dev/pandas/blob/d43ac979fdf3038c713de93c7b27b4140be6c9c5/pandas/io/formats/html.py#L307-L309

There is currently no test to explicitly cover this example. so i think the best way forward is to fully parameterize the truncation tests in line with the parametrized basic_alignment tests for added assurance.

i'll make `show_col_idx_names` a class property for clarity and add a note to refactor and 'inherit' from `DataFrameFormatter` class. inherit quoted since `HTMLFormatter` class is not directly inherited from `DataFrameFormatter`. in the first refactor just use mock inheritence like..

https://github.com/pandas-dev/pandas/blob/d43ac979fdf3038c713de93c7b27b4140be6c9c5/pandas/io/formats/html.py#L46-L48